File: trees/easy-subtree-of-another-tree/easy-subtree-of-another-tree-rs/src/sol1.rs
use std::cell::RefCell;
use std::rc::Rc;

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

type NodeRef = Rc<RefCell<TreeNode>>;

// 8ms, 2.2MB
pub fn is_subtree(root: Option<NodeRef>, sub_root: Option<NodeRef>) -> bool {
    is_subtree_recursive(&root, &sub_root)
}

pub fn is_subtree_recursive(root: &Option<NodeRef>, sub_root: &Option<NodeRef>) -> bool {
    match (root, sub_root) {
        (None, None) => true,
        (Some(r), Some(s)) => {
            if is_subtree_recursive(root, sub_root) {
                return true;
            }

            let r = r.borrow();

            is_subtree_recursive(&r.left, sub_root) || is_subtree_recursive(&r.right, sub_root)
        }
        _ => false,
    }
}

pub fn is_same_tree(p: &Option<NodeRef>, q: &Option<NodeRef>) -> bool {
    match (p, q) {
        (None, None) => true,
        (Some(p), Some(q)) => {
            let (p, q) = (p.borrow(), q.borrow());

            if p.val != q.val {
                return false;
            }

            is_same_tree(&p.left, &q.left) && is_same_tree(&p.right, &q.right)
        }
        _ => false,
    }
}

File: trees/medium-lowest-common-ancestor-of-a-binary-search-tree/medium-lowest-common-ancestor-of-a-binary-search-tree-go/sol1.go
package mediumlowestcommonancestorofabinarysearchtree

// Definition for a binary tree node
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 20ms, 6.9MB
func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	// traverse tree
	// set up queue
	var queue []*TreeNode
	queue = append(queue, root)

	var lca *TreeNode
	// bfs
	for {
		l := len(queue)
		if l == 0 {
			break
		}

		// pop from queue
		for i := 0; i < l; i++ {
			pop := queue[0]
			queue = queue[1:]

			// not a common ancestor
			if !bfsIsContains(pop, p, q) {
				continue
			}

			lca = pop

			if pop.Left != nil {
				queue = append(queue, pop.Left)
			}
			if pop.Right != nil {
				queue = append(queue, pop.Right)
			}

		}
	}

	return lca
}

func bfsIsContains(n, p, q *TreeNode) bool {
	if n == nil || p == nil || q == nil {
		return false
	}

	var (
		isPExist, isQExist = false, false
	)

	if n.Val == p.Val {
		isPExist = true
	}
	if n.Val == q.Val {
		isQExist = true
	}

	// set up queue
	var queue []*TreeNode
	queue = append(queue, n)

	// bfs
	for {
		l := len(queue)
		if l == 0 {
			break
		}

		// pop from queue
		for i := 0; i < l; i++ {
			pop := queue[0]
			queue = queue[1:]

			if pop.Val == p.Val {
				isPExist = true
			}
			if pop.Val == q.Val {
				isQExist = true
			}

			if isPExist && isQExist {
				return true
			}

			if pop.Left != nil {
				queue = append(queue, pop.Left)
			}
			if pop.Right != nil {
				queue = append(queue, pop.Right)
			}
		}
	}

	return isPExist && isQExist
}
